### PR TITLE
Digipoort

### PR DESCRIPTION
**Pull request by angela2389**
Hi Microsoft team,

Is it possible to add Digipoort as a module? 

This Pull Request is not yet complete, before I make the code allign with the Microsoft Extensions requirements I would like to know if you would be willing to add it first.

Also, can you advise if the codeunit + dotnet.al are in the correct folder? And do you have digipoort certificates available to run the test codeunit? Or should I make a test codeunit which will return an error from Digipoort with 'incorrect certificate'?

Thanks in advance, 
Angela

### CODE REVIEW

**Comment by JesperSchulz** [comment]
Hi Angela!

Thank you for your suggestion! We were wondering, if your PR isn't covered by the already existing [NL Tax Declaration Add-On](https://github.com/microsoft/ALAppExtensions/tree/master/AddOns/NLDigitalTaxDeclaration/app) here on GitHub? If that one doesn't cover you needs, could you maybe add a PR to enhance that add-on extension or let us know, what we've missed?

**Comment by angela2389** [comment]
Hi Jesper,

No, unfortunately it is not. The Add-On you are referring to is made voor VAT report, but I want to submit Payroll Tax. In C/AL we already have a codeunit to submit the Payroll Tax by our own Codeunit.

We also have our own codeunit to create the XML, we don't need an Add On for this, as we use all the XML variables available in AL. This is definitely not same/similiar to the XML for the VAT report.

Therefore I only need a codeunit to submit the XML and receive response messages from Digipoort, which require the digipoort variables which are not standaard in AL.

**Comment by angela2389** [comment]
Hi @JesperSchulz,

In the meanwhile I have looked into "Elec Tax Declaration Mgt" Codeunit and see that in BC16 the function SubmitDeclaration has been added. This is the function that is called from the Add On you mentioned earlier and which is similiar as my Pull Request.

I guess I cannot amend this Codeunit myself as it's in the Base App. Is this something you could arrange? Can the function I created in this Pull Request be added to that codeunit? And can this function be called from a new Add-On Codeunit?

**Comment by JesperSchulz** [comment]
That is definitely something we could look into! As I am no expert on local features, I'll have to forward your request to a team which is. Please allow me to present your suggestion to them and get back to you :-)

**Comment by JesperSchulz** [comment]
About time I reply to this open case. Sorry about the delay on this one as well! It took me some time to figure out, what options we've got. The way I see it, we've got two:

1) [optimal] You create a [System Application Module](https://github.com/microsoft/ALAppExtensions/tree/master/Modules/System), which takes care of generic communication with Digipoort. Once completed, I will make sure we uptake that module in our Add-On extension ourselves. System Application modules cannot take dependencies on any Base Application concepts and will hence by definition become generically useable.

2) We collaborate on tweaking the existing solution in the Base Application to suite your needs. However, collaboration will be more difficult, as I cannot open source parts of the Base Application. Also, a generic Digipoort communication API does not belong in the Base Application. It really is a poster child for a System Application module.

What do you think? Could we make option 1 work?

**Comment by angela2389** [comment]
Hi @JesperSchulz,

I will try to make a Pull Request for option 1, but there is still one problem remaining:
The Certificates Codeunit is 'OnPrem'. How can we make Digipoort work in the cloud? 

All this effort of making DotNet variables available in System Application Module is to make extensions cloud ready.

**Comment by JesperSchulz** [comment]
The scope attribute got some unfortunate option names. While we refer to them as OnPrem and Cloud, it does not mean that the code only runs OnPrem or in the cloud. OnPrem merely means, that cloud **apps** cannot access the code. The System Application and Base Application can! If you would make a system application module, you could access the code, given that the code would be in the system application as well (the system app cannot refer to the Base Application).
Please allow me to reach out to one of our security experts to hear what our options are, to move certificate management to the System Application or if we maybe already have something that would work.

**Comment by JesperSchulz** [comment]
Turns out, our System Application already has most of the certificate handling you will need (if not all). Take a look [here](https://github.com/microsoft/ALAppExtensions/tree/master/Modules/System/Cryptography%20Management#x509certificate2-codeunit-1286). That should be enough to create a "general purpose Digipoort communication" module?

**Comment by angela2389** [comment]
Hi @JesperSchulz, yes, I was able to implement the certifcate code within the Pull Request and tested the code within our application and it worked :). Can you please review my new Pull Request?

Please note: So far this is only the submission, within a few weeks I will also submit a codeunit to retrieve the response messages from digipoort.

**Comment by JesperSchulz** [comment]
We will be looking into this PR this week :-)

File: AddOns/Digipoort/app/Digipoort.al [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// ------------------------------------------------------------------------------------------------
+codeunit 50102 "Digipoort Payroll Tax"
+{
+    trigger OnRun()
+    var
+    begin
+    end;
+
+    procedure FuncSubmitPayrollTaxDeclaration(XmlContent: Text; ClientCertificateCode: Code[20]; ServiceCertificateCode: Code[20]; MessageType: Text; Reference: Text; RequestUrl: Text; VATReg: Text; Var MessageID: Text)
```
Comment (mazhelez) [inline AddOns/Digipoort/app/Digipoort.al:12]: No need for the `Func` preffix.

File: AddOns/Digipoort/app/Digipoort.al [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        UTF8Encoding: DotNet UTF8Encoding;
+        Window: Dialog;
+        WindowStatusBuildingMsg: Label 'Building document';
+        WindowStatusSendMsg: Label 'Transmitting document';
+        WindowStatusSaveMsg: Label 'Saving document ID';
+        WindowStatusMsg: Label 'Submitting Electronic Tax Declaration...\\Status          #1##################';
+        SubmitSuccessMsg: Label 'Declaration %1 was submitted successfully.';
+        SubmitErr: Label 'Submission of declaration %1 failed with error code %2 and the following message: \\%3';
+    begin
+        Window.Open(WindowStatusMsg);
```
Comment (mazhelez) [inline AddOns/Digipoort/app/Digipoort.al:50]: Is this always going to happen on the foreground? What if GUI isn't allowed, e.g. on background sessions?

File: AddOns/Digipoort/app/Digipoort.al [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        Fault := Fault.foutType;
+
+        UTF8Encoding := UTF8Encoding.UTF8Encoding;
+
+        with Identity do begin
+            nummer := VATReg;
+            type := 'LHnr';
+        end;
+
+        with Content do begin
```
Comment (mazhelez) [inline AddOns/Digipoort/app/Digipoort.al:65]: Avoid using the `with` statement. It's going to be deprecated in the future :)

File: AddOns/Digipoort/app/Digipoort.al [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        Response := DeliveryService.Deliver(Request,
+            RequestUrl,
+            ClientCertificateBase64,
+            DotNetSecureString,
+            ServiceCertificateBase64,
+            30);
+
+        Fault := Response.statusFoutcode;
+        if Fault.foutcode <> '' then
+            Error(StrSubstNo(SubmitErr, Reference, Fault.foutcode, Fault.foutbeschrijving));
```
Comment (mazhelez) [inline AddOns/Digipoort/app/Digipoort.al:93]: No need to use `StrSubstNo` in the `Error` function :)

File: AddOns/Digipoort/dotnet.al [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        Version = '15.0.0.0';
+        Culture = 'neutral';
+        PublicKeyToken = '31bf3856ad364e35';
+
+        type("Microsoft.Dynamics.NL.DigipoortServices"; "digipoortServices")
+        {
+        }
+
+        type("Digipoort.AanleverService.aanleverRequest"; "aanleverRequest")
```
Comment (mazhelez) [inline AddOns/Digipoort/dotnet.al:18]: Even though the .Net classes are in Dutch, it's better to give them corresponding English names :)

File: Modules/System Tests/Digipoort/DigipoortTest.Codeunit.al [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        ElecDeclarationSetup.Get();
+        VarUrl := ElecDeclarationSetup."Digipoort Delivery URL";
+        ServiceCertificate := ElecDeclarationSetup."Service Certificate Code";
+        ClientCertificate := ElecDeclarationSetup."Client Certificate Code";
+
+        // [WHEN] Send Xml to Digipoort
+        Digipoort.FuncSubmitPayrollTaxDeclaration(GetXml(), ClientCertificate, ServiceCertificate, VarMessageType, VarReference, VarUrl, VATRegNo, MessageID);
+
+        // [THEN] Verify Result 
+        IF MessageID = '' THEN
```
Comment (mazhelez) [inline Modules/System Tests/Digipoort/DigipoortTest.Codeunit.al:40]: Use `Assert` is tests.

File: AddOns/Digipoort/dotnet.al (hunk 5 above)
Comment (angela2389) [reply AddOns/Digipoort/dotnet.al:18]: Hi Maria, these dotnet variables actually got implemented in BC16 base application (including the dutch names ;)), therefore I don't think the dotnet.al is needed anymore. I will remove it.

File: AddOns/Digipoort/app/Digipoort.al (hunk 2 above)
Comment (angela2389) [reply AddOns/Digipoort/app/Digipoort.al:50]: It should always happen in the foreground, but I will add GUIALLOWED checks around it.

**Comment by angela2389** [comment]
> The scope attribute got some unfortunate option names. While we refer to them as OnPrem and Cloud, it does not mean that the code only runs OnPrem or in the cloud. OnPrem merely means, that cloud **apps** cannot access the code. The System Application and Base Application can! If you would make a system application module, you could access the code, given that the code would be in the system application as well (the system app cannot refer to the Base Application).
> Please allow me to reach out to one of our security experts to hear what our options are, to move certificate management to the System Application or if we maybe already have something that would work.

Hi @JesperSchulz and @mazhelez I used the IsolatedCertificate functions in the end as it also contains the DotNetSecureString which is needed to send the certifcate. I'm aware of the other X509 certificate functions as well (I use them to submit pension ;)).

**Comment by angela2389** [comment]
> Nice work! Still a lot of work left though!
> 
> Have a look at @JesperSchulz's comment and try to implement it :)

Hi @mazhelez Not too sure which comment you refer too, but I moved it to Modules and added app.json and the readme file. Hope this is fine now.

File: Modules/System/Tax Management/src/Digipoort.Codeunit.al [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    /// <param name="XmlContent">The payroll tax declartion in XML format</param>
+    /// <param name="ClientCertificateCode">The client certificate code from the certificate stored in Isolated Certificate</param>
+    /// <param name="ServiceCertificateCode">The service certificate code from the certificate stored in Isolated Certificate</param>
+    /// <param name="MessageType">The MessageType of the message</param>
+    /// <param name="Reference">The delivery reference to be send to Digipoort</param>
+    /// <param name="RequestUrl">The url the tax declaration needs to be send to</param>
+    /// <param name="VATReg">The VAT Registration Number to be send to Digipoort</param>
+    /// <param name="MessageID">The message ID to be received back from Digipoort</param>
+
+    procedure SubmitPayrollTaxDeclaration(XmlContent: Text; ClientCertificateCode: Code[20]; ServiceCertificateCode: Code[20]; MessageType: Text; Reference: Text; RequestUrl: Text; VATReg: Text): Text
```
Comment (JesperSchulz) [inline Modules/System/Tax Management/src/Digipoort.Codeunit.al:31]: Could we make this "general purpose". Could this submit procedure not be used to send any content, or is this for payroll only?

File: Modules/System/Tax Management/src/app.json [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+                   "publisher": "Microsoft",
+                   "version": "16.0.0.0"}
+                     ],
+    "screenshots":  [
+
+                    ],
+    "platform":  "15.0.0.0",
+    "idRanges":  [
+                     {
+                         "from":  50102, //Microsoft team?
```
Comment (JesperSchulz) [inline Modules/System/Tax Management/src/app.json:25]: We are going to update the IDs, when you're done :-)

File: Modules/System/Tax Management/src/app.json [hunk 9]
```diff
@@ -0,0 +1,31 @@
+{
+    "id":  "93f93dda-3177-4de6-9f6e-b38103809c6d",
+    "name":  "Tax Management",
```
Comment (JesperSchulz) [inline Modules/System/Tax Management/src/app.json:3]: I think this name is too generic, as this only works in the Netherlands. Let's work on the name. Dutch Tax Management? But this isn't really tax managment either, is it? It's the Digipoort submit part. Maybe Dutch Tax Submission? Naming discussions can be fun :-D

Comment (angela2389) [reply Modules/System/Tax Management/src/app.json:3]: Then we might as well just call it Digipoort? ;) I thought you might want to re-use it for other countries.

File: Modules/System/Tax Management/src/Digipoort.Codeunit.al (hunk 7 above)
Comment (angela2389) [reply Modules/System/Tax Management/src/Digipoort.Codeunit.al]: Well in that case I added one more parameter, it should be usable for other content now as well.

**Comment by angela2389** [comment]
@JesperSchulz and @mazhelez I will come with a PR for for the responses today as well.

**Comment by JesperSchulz** [comment]
I really appreciate the effort, Angela! Let me know, if you'd like a short call to go over your PR and the open suggestions, or if this format is ok for you? I realize it's quite the learning curve to submit your first full module, but I'd like it to be as "painless" for you as possible :-)

**Comment by angela2389** [comment]
Hi Jesper, a call sounds good! Can you email me at angela_2389@hotmail.com to discuss the details further?

**Comment by JesperSchulz** [comment]
Sorry, didn't read your message until now :-) You should have a message in your inbox now.

File: Modules/System/Digipoort/src/app.json [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    "brief":  "Provides tax helper functions",
+    "description":  "Provides functions to submit tax and receive response messages",
+    "version":  "16.0.0.0",
+    "privacyStatement":  "https://go.microsoft.com/fwlink/?linkid=724009",
+    "EULA":  "https://go.microsoft.com/fwlink/?linkid=2009120",
+    "help":  "https://go.microsoft.com/fwlink/?linkid=2103698",
+    "url":  "https://go.microsoft.com/fwlink/?linkid=724011",
+    "logo":  "",
+    "dependencies":  [
+                  {"appId": "437dbf0e-84ff-417a-965d-ed2bb9650972",
```
Comment (JesperSchulz) [inline Modules/System/Digipoort/src/app.json:14]: This dependency is not possible. The System Application cannot reference the Base Application.

**Comment by bc-ghost** [comment]
Thanks for submitting this issue. We will publish a fix for it in the next minor update for release 16.

**Comment by bc-ghost** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. We will update this issue with information about availability.

**Comment by darjoo** [comment]
The changes have been implemented in the NL BaseApp and will be available in the next minor release.

**Comment by bc-ghost** [comment]
Thanks for submitting this issue. We will publish a fix for it in the next minor update for release 15.

**Comment by darjoo** [comment]
This has been backported to 15 as well.

**Pull request closed without merging**